Add fingerprint processor (#7631)
opensearch-project/documentation-website: _ingest-pipelines/processors/fingerprint.md

---
layout: default
title: Fingerprint
parent: Ingest processors
nav_order: 55
---

> # Fingerprint processor

Reviewer: @gaobinlong Thank you for drafting this content. Doc review is completed. I made some edits for clarity. Please approve at your availability, and then I'll move this PR to editorial. Thank you.

Author: @vagimeli Thanks for your review, but I cannot approve this PR because I'm the author.

Introduced 2.16
{: .label .label-purple }

> The `fingerprint` processor is used to generate a hash value for the specified fields or all fields in a document. The hash value can be used to deduplicate documents within a index and collapse search results.

Reviewer: ```suggestion
The `fingerprint` processor is used to generate a hash value for the specified fields or all fields in a document. The hash value can be used to deduplicate documents within an index and collapse search results.
```

> The `fingerprint` processor is used to generate a hash value for the specified fields or all fields in a document. The hash value can be used to deduplicate documents within an index and collapse search results.

Reviewer 2: ```suggestion
The `fingerprint` processor is used to generate a hash value for either certain specified fields or all fields in a document. The hash value can be used to deduplicate documents within an index and collapse search results.
```

> For each field, the field name, length of the field value, and the field value itself are concatenated and separated by the pipe character `|`. For example, if the field name is `field1` and the value is `value1`, then the concatenated string would be`|field1|3:value1|field2|10:value2|`. For object fields, the field name is flattened by joining the nested field names with a period `.`. For instance, if the object field is `root_field` with a sub-field `sub_field1` having the value `value1` and another sub-field `sub_field2` with the value `value2`, the concatenated string would be `|root_field.sub_field1|1:value1|root_field.sub_field2|100:value2|`.

Reviewer: ```suggestion
For each field, the field name, length of the field value, and the field value itself are concatenated and separated by the pipe character `|`. For example, if the field name is `field1` and the value is `value1`, then the concatenated string would be`|field1|3:value1|field2|10:value2|`. For object fields, the field name is flattened by joining the nested field names with a period `.`. For instance, if the object field is `root_field` with a sub-field `sub_field1` having the value `value1` and another sub-field `sub_field2` with the value `value2`, then the concatenated string would be `|root_field.sub_field1|1:value1|root_field.sub_field2|100:value2|`.
```

> For each field, the field name, length of the field value, and the field value itself are concatenated and separated by the pipe character `|`. For example, if the field name is `field1` and the value is `value1`, then the concatenated string would be`|field1|3:value1|field2|10:value2|`. For object fields, the field name is flattened by joining the nested field names with a period `.`. For instance, if the object field is `root_field` with a sub-field `sub_field1` having the value `value1` and another sub-field `sub_field2` with the value `value2`, then the concatenated string would be `|root_field.sub_field1|1:value1|root_field.sub_field2|100:value2|`.

Reviewer 2: ```suggestion
For each field, the field name, the length of the field value, and the field value itself are concatenated and separated by the pipe character `|`. For example, if the field name is `field1` and the value is `value1`, then the concatenated string would be `|field1|3:value1|field2|10:value2|`. For object fields, the field name is flattened by joining the nested field names with a period `.`. For instance, if the object field is `root_field` with a sub-field `sub_field1` having the value `value1` and another sub-field `sub_field2` with the value `value2`, then the concatenated string would be `|root_field.sub_field1|1:value1|root_field.sub_field2|100:value2|`.
```

> The following is the `fingerprint` processor syntax:

Reviewer 2: ```suggestion
The following is the syntax for the `fingerprint` processor:
```

```json
{
  "community_id": {
    "fields": ["foo", "bar"],
    "target_field": "fingerprint",
    "hash_method": "SHA-1@2.16.0"
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `fingerprint` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
`fields`  | Optional  | A list of fields used to generate a hash value.  |

> `exclude_fields`  | Optional  | Specifies the fields to be excluded from hash value generation. It is mutually exclusive with the `fields` parameter; if both `exclude_fields` and `fields` are empty or null, all fields are included in the hash value calculation. |

Reviewer 2: ```suggestion
`exclude_fields`  | Optional  | Specifies the fields to be excluded from hash value generation. It is mutually exclusive with the `fields` parameter; if both `exclude_fields` and `fields` are empty or null, then all fields are included in the hash value calculation. |
```

`hash_method`  | Optional  | Specifies the hashing algorithm to be used, with options being `MD5@2.16.0`, `SHA-1@2.16.0`, `SHA-256@2.16.0`, or `SHA3-256@2.16.0`. Default is `SHA-1@2.16.0`. The version number is appended to ensure consistent hashing across OpenSearch versions, and new versions will support new hash methods. |

> `target_field`  | Optional  | Specifies the name of the field where the generated hash value will be stored. If not provided, the hash value is stored in the `fingerprint` field by default. |

Reviewer 2: ```suggestion
`target_field`  | Optional  | Specifies the name of the field in which the generated hash value will be stored. If not provided, then the hash value is stored in the `fingerprint` field by default. |
```
